### PR TITLE
📝 (#9): Update README.md, LICENSE, and package.json

### PR DESCRIPTION
**Pull request by jeanaica**

### CODE REVIEW

**Comment by jeanaica** [comment]
Based on the patch provided, it appears that there are no changes made to the original source file. The patch simply adds the entire content of the original source file. 

If there were any specific changes made to the source file, please provide the diff or patch that highlights those changes so that I can provide a more meaningful review and suggestions for improvement.

**Comment by jeanaica** [comment]
Certainly! Let's review the changes made between the source files and identify any differences or duplications.

Based on the patch, here are the changes made:

1. A heading `# Life in Peach (CMS API)` is added at the beginning of the file.
2. The project description is added, explaining that it is a CMS API built on Firebase and Firestore.
3. The section "Dependencies" is added, listing the primary packages and technologies used in the project.
4. The section "Setting Up" is added, providing instructions for local setup.
5. The section "Environment Variables" is added, explaining how to configure the environment variables.
6. The section "Installing Dependencies" is added, indicating how to install all project dependencies.
7. The section "Building and Running the Project" is added, explaining how to build and start the project.
8. The section "Scripts" is added, listing useful NPM scripts defined in the `package.json` file.

Overall, the changes made in the patch provide additional context and instructions for setting up and running the project.

As for suggestions for improvement, the patch seems to cover the necessary information. However, one improvement could be to provide more specific version requirements for the dependencies, especially for Node.js and Firebase CLI. This would ensure that users have the correct versions installed.

Regarding duplications, I don't see any duplications in the patch. The changes provide valuable information without repeating or conflicting with existing content.

Overall, the patch seems to enhance the documentation and provide clear instructions for getting started with the project.

**Comment by jeanaica** [comment]
Certainly! Here's a review of the changes made:

```diff
 "scripts": {
    "lint": "eslint --ext .js,.ts .",
    "postbundle": "copy-and-watch src/**/*.graphql lib/",
    "postbundle:watch": "copy-and-watch --watch src/**/*.graphql lib/",
    "build:watch": "tsc -w && npm run postbundle:watch",
    "build": "tsc && npm run postbundle",
-    "serve": "npm run build && firebase emulators:start --only functions",
-    "shell": "npm run build && firebase functions:shell",
-    "start": "npm run shell",
-    "deploy": "npm run build && firebase deploy --only functions",
+    "serve": "npm run build && firebase emulators:start --only functions", 
+    "shell": "npm run build && firebase functions:shell", 
+    "start": "npm run shell", 
+    "deploy": "npm run fb:prod && npm run build && firebase deploy --only functions", 
+    "deploy:staging": "npm run fb:dev && npm run build && firebase deploy --only functions", 
     "logs": "firebase functions:log",
-    "dev": "npm run build && firebase emulators:start",
+    "dev": "npm run build && firebase emulators:start", 
+    "fb:dev": "firebase use staging", 
+    "fb:prod": "firebase use prod"
   },
```

- The `deploy:staging` script has been added. This script is used to deploy to a staging environment.
- The `fb:dev` and `fb:prod` scripts have been added. These scripts are used to switch between staging and production Firebase projects.

Overall, the changes seem to add support for deploying to both staging and production environments using different Firebase projects.

Regarding possible improvements or duplications, I don't see any immediate concerns or duplications in the changes made in this patch. However, it's worth noting that there is some duplication in the `deploy` script where both `npm run build` and `firebase deploy --only functions` are called in both `deploy` and `deploy:staging` scripts. Consider refactoring these scripts to avoid duplication if it becomes a concern in the future.

If you have any further questions or need additional assistance, feel free to ask!

**Comment by jeanaica** [comment]
Sure, I'd be happy to review the changes made in the patch. 

Upon comparing the patch with the source code, I see that the patch adds the entire content of the original license without any changes. It simply adds the license text as is. 

In terms of improvements or suggestions, since the patch is adding a license to the code, it is generally recommended to have a separate file for the license rather than embedding it directly in the code. This follows best practices and makes it easier for others to find and reference the license. You can create a file named `LICENSE` in the root directory of your project and place the license text there. Then you can update your patch to reference the license file.

Here's an example of how you can update the patch:

```diff
@@ -0,0 +1,3 @@
+/* The MIT License (MIT) */
+
+/* Add a comment here to indicate that the license file is present. */
```

Then in your project's root directory, create a file named `LICENSE` and place the license text there.

By doing this, your project will have a separate license file that can be easily identified, and it adheres to the best practices followed by most software projects.

As for duplications, I don't see any duplications in the provided code. The license text is added only once.

I hope this helps! Let me know if you have any further questions.

**Comment by jeanaica** [comment]
Certainly! Let's review the changes made in the patch and see if there are any improvements or duplications.

Here are the changes made in the patch:

```diff
+ # Life in Peach (CMS API)
+
+ This project provides a Content Management System (CMS) API built on top of Firebase and Firestore. It allows you to manage and deliver digital content to various types of applications.
+
+ ## Dependencies
+
+ This project uses various packages and technologies. The primary ones are Firebase, Firestore, Express, Apollo Server Express, and GraphQL. See `package.json` for a full list of dependencies and devDependencies.
+
+ ## Setting Up
+
+ ### Pre-requisites
+
+ To run this project locally, you need:
+
+ - Node.js 16
+ - NPM
+ - Firebase CLI
+ - Access to a Firebase project
+
+ ### Environment Variables
+
+ Copy the `.env.example` file and rename it to `.env`. Replace the placeholder values with your actual configuration:
 +
+ ```
+ CONFIG_TOKEN="your-firebase-token"
+ CONFIG_ENV="your-environment"
+ CONFIG_PROJECT_ID="your-firebase-project-id"```
+

- ## Installing Dependencies

- In the project directory, run `npm install` to install all dependencies.
+ ## Installing Dependencies
+
+ In the project directory, run `npm install` to install all dependencies.

- ## Building and Running the Project

- To build the project, run `npm run build`. This will compile the TypeScript files and generate JavaScript files in the `lib/` directory.

- To start the project, run `npm run dev`. This will start the Firebase emulator.
+ ## Building and Running the Project
+
+ To build the project, run `npm run build`. This will compile the TypeScript files and generate JavaScript files in the `lib/` directory.

+ To start the project, run `npm run dev`. This will start the Firebase emulator.

- ## Scripts

- Here are some useful NPM scripts defined in `package.json`:

- - `npm run build`: Build the project.
- - `npm run serve`: Build and start the project and run emulator.
- - `npm run deploy`: Deploy the functions to Firebase to prod.
- - `npm run deploy:staging`: Deploy the functions to Firebase to staging.
- - `npm run logs`: View the Firebase functions logs.
- - `npm run fb:dev`: Switch Firebase to the dev project.
- - `npm run fb:prod`: Switch Firebase to the prod project.

+ ## Scripts
+
+ Here are some useful NPM scripts defined in `package.json`:
+
+ - `npm run build`: Build the project.
+ - `npm run serve`: Build and start the project and run emulator.
+ - `npm run deploy`: Deploy the functions to Firebase to prod.
+ - `npm run deploy:staging`: Deploy the functions to Firebase to staging.
+ - `npm run logs`: View the Firebase functions logs.
+ - `npm run fb:dev`: Switch Firebase to the dev project.
+ - `npm run fb:prod`: Switch Firebase to the prod project.
```

Upon reviewing the changes, it seems that the patch adds the project title, an overview, and instructions for setting up the environment variables. Additionally, the bullet points under "Installing Dependencies" and "Building and Running the Project" have been formatted differently. In the original source, each point had a separate line, while in the patch, they are all in one line.

As for improvements, it would be helpful to provide more details or links about how to obtain a Firebase token, the specific environment values, and the Firebase project ID. This would help users set up their configuration more easily.

Regarding duplications, there doesn't seem to be any significant duplication in the changed code. Some minor duplication is present in the script descriptions, but it's minimal and doesn't impact the functionality.

Overall, the changes made in the patch provide a clearer overview of the project and instructions for setting up the environment. However, providing more specific information and improving the formatting further would enhance the patch.

**Comment by jeanaica** [comment]
Certainly! Let's review the changes made in the patch and identify any potential improvements or duplications.

```diff
   "scripts": {
     "lint": "eslint --ext .js,.ts .",
     "postbundle": "copy-and-watch src/**/*.graphql lib/",
     "postbundle:watch": "copy-and-watch --watch src/**/*.graphql lib/",
     "build:watch": "tsc -w && npm run postbundle:watch",
-    "build": "tsc && npm run postbundle",
+    "build": "tsc && npm run postbundle",
     "serve": "npm run build && firebase emulators:start --only functions",
     "shell": "npm run build && firebase functions:shell",
     "start": "npm run shell",
-    "deploy": "npm run build && firebase deploy --only functions",
+    "deploy": "npm run build && firebase deploy --only functions",
     "deploy:staging": "npm run fb:dev && npm run deploy",
     "deploy:prod": "npm run fb:prod && npm run deploy",
     "logs": "firebase functions:log",
-    "dev": "npm run build && firebase emulators:start"
+    "dev": "npm run build && firebase emulators:start",
+    "fb:dev": "firebase use staging",
+    "fb:prod": "firebase use prod"
   },
```

1. The changes made in the `scripts` section are as follows:
   - Added a new script `"deploy:staging"` which executes the commands in sequence `npm run fb:dev` and `npm run deploy`.
   - Added a new script `"deploy:prod"` which executes the commands in sequence `npm run fb:prod` and `npm run deploy`.
   - Added a new script `"fb:dev"` which sets the Firebase project configuration to staging.
   - Added a new script `"fb:prod"` which sets the Firebase project configuration to prod.

In terms of potential improvements and duplications, there are a few points to consider:

1. Duplicated code: The scripts `"deploy:staging"` and `"deploy:prod"` execute the command `npm run deploy` after executing other commands. Since the `"deploy"` script already includes `npm run build`, it may be more efficient to directly call `npm run deploy` from `"deploy:staging"` and `"deploy:prod"`, without duplicating the `npm run build` step. This would reduce duplication and improve maintainability.

2. Consistency: The scripts `"deploy:staging"` and `"deploy:prod"` prepend `"fb:dev"` and `"fb:prod"` respectively before calling `"deploy"`. Consider implementing a more consistent approach where `"deploy"` script itself sets the Firebase project configuration based on an environment variable or a command-line argument. This would simplify the scripts and avoid duplication.

Here's an updated version of the patch, addressing the above points:

```diff
   "scripts": {
     "lint": "eslint --ext .js,.ts .",
     "postbundle": "copy-and-watch src/**/*.graphql lib/",
     "postbundle:watch": "copy-and-watch --watch src/**/*.graphql lib/",
     "build:watch": "tsc -w && npm run postbundle:watch",
     "build": "tsc && npm run postbundle",
     "serve": "npm run build && firebase emulators:start --only functions",
     "shell": "npm run build && firebase functions:shell",
     "start": "npm run shell",
     "deploy": "npm run build && firebase deploy --only functions",
-    "deploy:staging": "npm run fb:dev && npm run deploy",
-    "deploy:prod": "npm run fb:prod && npm run deploy",
     "logs": "firebase functions:log",
     "dev": "npm run build && firebase emulators:start",
+    "deploy:staging": "firebase use staging && npm run deploy",
+    "deploy:prod": "firebase use prod && npm run deploy"
   },
   "engines": {
     "node": "16"
   }
```

With these changes, the code becomes more streamlined and avoids duplications while maintaining consistency.

Please let me know if there's anything else I can help you with!

**Comment by jeanaica** [comment]
Sure, let's review the changes made in the patch.

The patch appears to be adding the entire content of the original source file. It does not introduce any changes or modifications to the original license text. Therefore, there are no differences between the patch and the original source.

As for suggestions to improve the patch or identify duplications, since the patch is adding the same content as the original source file, it can be considered unnecessary and redundant. If this patch is intended as an update to the license, it should contain the actual changes or modifications to the license text.

In summary, the patch does not introduce any changes or improvements and can be considered redundant. It would be more meaningful to include only the relevant modifications to the license text in the patch, if any.

**Comment by jeanaica** [comment]
Certainly! Let's review the changes made in the patch and suggest any improvements:

The changes made in the patch include adding a project description, listing dependencies, providing setup instructions, describing environment variables, and documenting npm scripts.

Overall, the changes made in the patch provide clear and concise instructions for setting up and running the project. Here are some suggestions for improvement:

1. Add a brief introduction: It would be helpful to provide a brief description of the project and its purpose at the beginning of the README file.

2. Organize the sections: Consider grouping related information under appropriate headings to improve readability and make it easier for users to find what they need. For example, you can group the setup instructions and environment variables under a single heading.

3. Provide more details about dependencies: While mentioning the primary dependencies is good, it would be helpful to provide specific versions or ranges for each dependency. This helps ensure compatibility and allows users to install specific versions if needed.

4. Add code snippets and examples: Consider including code snippets or examples to illustrate how to use certain features or interact with the API. This can help users understand the functionality better and see the API in action.

Regarding duplications, there are no apparent duplications in the provided patch. However, it's always a good practice to review the entire codebase and ensure there are no redundant or duplicated code blocks.

Here's an updated version of the patch with the suggested improvements:

```markdown
# Life in Peach (CMS API)

This project is a Content Management System (CMS) API built on top of Firebase and Firestore. It allows you to manage and deliver digital content to various types of applications.

## Introduction

[Provide a brief introduction to the project and its purpose.]

## Dependencies

This project uses the following dependencies (specific versions are recommended):

- Firebase
- Firestore
- Express
- Apollo Server Express
- GraphQL

See `package.json` for a full list of dependencies and devDependencies.

## Setup

### Prerequisites

To run this project locally, you'll need:

- Node.js 16
- NPM
- Firebase CLI
- Access to a Firebase project

### Environment Variables

1. Copy the `.env.example` file and rename it to `.env`.
2. Replace the placeholder values with your actual configuration:

```dotenv
CONFIG_TOKEN="your-firebase-token"
CONFIG_ENV="your-environment"
CONFIG_PROJECT_ID="your-firebase-project-id"
```

## Installation

1. In the project directory, run `npm install` to install all dependencies.

## Building and Running the Project

- To build the project, run `npm run build`. This will compile the TypeScript files and generate JavaScript files in the `lib/` directory.
- To start the project, run `npm run dev`. This will start the Firebase emulator.

## Scripts

Here are some useful NPM scripts defined in `package.json`:

- `npm run build`: Build the project.
- `npm run serve`: Build and start the project and run emulator.
- `npm run deploy`: Deploy the functions to Firebase to prod.
- `npm run deploy:staging`: Deploy the functions to Firebase to staging.
- `npm run logs`: View the Firebase functions logs.
- `npm run fb:dev`: Switch Firebase to the dev project.
- `npm run fb:prod`: Switch Firebase to the prod project.

```

I hope this review and the suggested improvements are helpful! Let me know if you have any further questions.

**Comment by jeanaica** [comment]
Certainly! Let's review the changes made in the patch and identify any issues or room for improvement.

Patch:

```diff
     "shell": "npm run build && firebase functions:shell",
     "start": "npm run shell",
     "deploy": "npm run build && firebase deploy --only functions",
+    "deploy:staging": "firebase use staging && npm run deploy",
+    "deploy:prod": "firebase use prod && npm run deploy",
     "logs": "firebase functions:log",
     "dev": "npm run build && firebase emulators:start"
   },
```

The patch adds two new scripts: `deploy:staging` and `deploy:prod`. These scripts are used to deploy the functions to the staging and production environments respectively.

Suggestions for improvement:

1. Duplications: The scripts `deploy:staging` and `deploy:prod` seem to be very similar to the existing `deploy` script. It would be better to avoid duplicating the deployment logic. One approach could be to modify the existing `deploy` script to accept an argument for the deployment environment.

For example:

```json
"deploy": "npm run build && firebase deploy --only functions",
```

Then, you can pass the environment as an argument:

```bash
npm run deploy -- --project staging
```

2. Environment variables: It's recommended to use environment variables for storing sensitive information, such as project IDs and API keys. Consider using a tool like `dotenv` or `env-cmd` to manage environment variables in different environments more securely.

3. Error handling: It's important to handle errors properly when executing these scripts. Currently, there is no error handling mechanism in place. Consider adding proper error handling and error messages to improve the developer experience.

4. Consistent naming: It's good to establish a naming convention for scripts to keep them consistent and easier to understand. For example, you could use `deploy:stage` instead of `deploy:staging` to align with the existing naming pattern.

Overall, the changes made in the patch are functional, but there are opportunities to improve code duplication and error handling.
